opentimelineview/timeline_widget.py

```diff
@@ -344,24 +353,262 @@ def counteract_zoom(self, zoom_level=1.0):
+
+    def set_time_space_callback (self, time_space) :
+        self._time_space = time_space
+        self.update_frame()
+
+    def mouseMoveEvent(self, mouse_event):
+        pos = self.mapToScene(mouse_event.pos())
+        self.setPos(QtCore.QPointF(pos.x(),
+            TIME_SLIDER_HEIGHT - MARKER_SIZE))
+        self.update_frame()
+
+        super(Ruler, self).mouseMoveEvent(mouse_event)
+     
+    def hoverEnterEvent (self,event):
+        self.setSelected(True)
+        super(Ruler, self).hoverEnterEvent(event)                
+        
+    def hoverLeaveEvent (self,event):
+        self.setSelected(False)
+        super(Ruler, self).hoverLeaveEvent(event)                
+
+    def init(self):
+        for item in self.composition.items():
+            if isinstance(item, Track):
+                frameNumber_tail = FrameNumber("",position=-1)
+                frameNumber_tail.setParentItem(self)
+                frameNumber_head = FrameNumber("",position=1)
+                frameNumber_head.setParentItem(self)
+                self.labels.append([item, frameNumber_tail, frameNumber_head])
+                frameNumber_tail.setY(item.pos().y())
+                frameNumber_head.setY(item.pos().y())
+
+        self.update_frame()
+
+    def setParentItem(self, timeSlider):
+        '''
+        subclass in order to add the rule to the timeSlider item.
+        '''
+        timeSlider.add_ruler(self)
+        super(Ruler, self).setParentItem(timeSlider)        
+
+    def map_to_time_space (self, item):     
+        '''
+        Temporary implementation.
+        @TODO: modify this function once Time Coordinates Spaces 
+        feature is implemented.
+        '''   
+        ratio = (self.x() - item.x()) / \
+                float(item.rect().width())
+        start_time = item.item.source_range.start_time.value
+        duration = item.item.source_range.duration.value
+
+        is_bounded = False
+        is_head = False
+        is_tail = False
+        f = "-?-"
+        f_nb = ratio * duration + start_time
+        if self._time_space == "frame_number":
+            f = ratio * duration
+        elif self._time_space == "media_space":
+            f = duration * ratio + start_time
+
+        if round(f_nb) >= start_time and round(f_nb) <= (start_time + duration) :
+            is_bounded = True
+        if round(f_nb) == start_time :
+            is_head = True
+        if round(f_nb) >= (start_time + duration) :
+            is_tail = True
+
+        bounded_data = namedtuple("bounded_data",["f","is_bounded","is_tail","is_head"])
+        return bounded_data(f, is_bounded, is_tail, is_head)
+
+    def update_frame(self):
+        for track, frameNumber_tail, frameNumber_head in self.labels:
+            f_tail = ""
+            f_head = ""
+            highlight_head = False
+            for item in track.childItems():
+                if isinstance(item, ClipItem) or isinstance(item, NestedItem):
```

Comment:
if you invert this if statement:
`if not (isinstance...)`
then you can dedent this block